unit/cq_test.c

```diff
@@ -0,0 +1,169 @@
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AWV
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <getopt.h>
+#include <poll.h>
+#include <time.h>
+#include <string.h>
+
+#include <rdma/fabric.h>
+#include <rdma/fi_domain.h>
+#include <rdma/fi_errno.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_cm.h>
+
+#include "unit_common.h"
+#include "shared.h"
+
+
+static char err_buf[512];
+
+static int
+create_cq(struct fid_cq **cq, size_t size, uint64_t flags,
+		enum fi_cq_format format, enum fi_wait_obj wait_obj)
+{
+	struct fi_cq_attr cq_attr;
+
+	memset(&cq_attr, 0, sizeof(cq_attr));
+	cq_attr.size = size;
+	cq_attr.flags = flags;
+	cq_attr.format = format;
+	cq_attr.wait_obj = wait_obj;
+
+	return fi_cq_open(domain, &cq_attr, cq, NULL);
+}
+
+/*
+ * Tests:
+ * - test open and close of CQ over a range of sizes
+ */
+static int
+cq_open_close()
+{
+	int i;
+	int ret;
+	int size;
+	int testret;
+	struct fid_cq *cq;
+
+	testret = FAIL;
+
+	for (i = -1; i < 17; ++i) {
+		size = (i < 0) ? 0 : 1 << i;
+		if (size > fi->domain_attr->cq_cnt)
+			break;
+
+		ret = create_cq(&cq, size, 0, FI_CQ_FORMAT_UNSPEC, FI_WAIT_UNSPEC);
+		if (ret != 0) {
+			sprintf(err_buf, "fi_cq_open(%d, 0, FI_CQ_FORMAT_UNSPEC, "
+					"FI_WAIT_UNSPEC) = %d, %s",
+					size, ret, fi_strerror(-ret));
+			goto fail;
+		}
+
+		ret = fi_close(&cq->fid);
+		if (ret != 0) {
+			sprintf(err_buf, "close(cq) = %d, %s", ret, fi_strerror(-ret));
+			goto fail;
+		}
+		cq = NULL;
+	}
+	testret = PASS;
+
+fail:
+	cq = NULL;
+	return TEST_RET_VAL(ret, testret);
+}
+
+struct test_entry test_array[] = {
+	TEST_ENTRY(cq_open_close),
+	{ NULL, "" }
+};
+
+int main(int argc, char **argv)
+{
+	int op, ret;
+	int failed;
+
+	hints = fi_allocinfo();
+	if (!hints)
+		exit(1);
+
+	while ((op = getopt(argc, argv, "f:a:")) != -1) {
+		switch (op) {
+		case 'a':
+			free(hints->fabric_attr->name);
+			hints->fabric_attr->name = strdup(optarg);
+			break;
+		case 'f':
+			free(hints->fabric_attr->prov_name);
+			hints->fabric_attr->prov_name = strdup(optarg);
+			break;
+		default:
+			printf("usage: %s\n", argv[0]);
```

Comment:
> Looks like none of the unit tests supports -h option. We could add one for all the unit tests. Also, could you please add a one-line description about the test before printing usage?

That's something I never noticed, good catch. I think it would be good for them to all support the `-h` parameter.